ruby_programming/advanced_ruby/blocks.md

```diff
@@ -0,0 +1,547 @@
+
+~~~ruby
+[1,2,3].each { |num| puts num }
+
+[1,2,3].each do |num|
+  puts num
+end
+~~~
+
+Although you are familiar with how to write blocks, you need to know how to write your own methods that accept blocks.
+
+### Yield
+
+`yield` is a keyword that can be called inside a method to relinquish execution to the accompanying block. Let's imagine you're writing a simple method for your library to allow users of that library to log some information to the terminal. The one key requirement you have is that users should be able to define how that block is presented. Sometimes they may just want to `puts` something, other times they may want to inspect something with `p`. If you tried to write this only using a method it would actually be quite difficult. You'd have to write up front all the possible options the user of your library might want to call and then create the docs to explain it. With blocks we can just relinquish control of the method to the block with yield and allow the user to define how they want to print it.
+
+~~~ruby
+def logger
+  yield
+end
+
+logger { puts 'hello from the block' }
+
+logger do
+  p [1,2,3]
+end
+~~~
+
+The example isn't that useful but hopefully you should see how blocks allow huge flexibility in how methods act upon data. Let's see if we can make it more interesting, but still not that useful. You now get given a new requirement that users want a method that allows them to write whatever they want, and it gets printed twice to the terminal. How might you handle this requirement? Well if I told you that you can call yield as many times as you want and each time it yields to the block in the same way you'd probably have a good idea.
+
+~~~ruby
+def double_vision
+  yield
+  yield
+end
+
+double_vision { puts "How many fingers am I holding up?" }
+~~~
+
+Millennial avocados Batman. That's cool.
+
+We mentioned earlier that blocks can accept arguments. But how do you pass them to the block? No problem! Just pass them as arguments to yield and they'll be passed to the blocks parameters. If you call yield more than once you can pass a different argument each time if you wanted to.
+
+Where might this be useful? Well think about each time you use `each`. You need to yield each member of the collection to the block. Say you're writing a banking api, it will allow other banks to use your library to handle all the banking matters. One such api is a method that prints all of the transactions to a statement. For our example we'll use the terminal, but it could be a spreadsheet or anything. One issue is that different banks may want to print the transaction with different formats. We can iterate through the transactions and for each one we can yield it to a block that the caller of your method attaches to the method call. They can define how the transactions will be printed to their statement and you can focus on delivering bug free banking transactions
```

Comment:
> writing a banking api

I don't think we teach anything about api's yet, I think this might cause some confusion. "writing a banking service" would maybe be clearer. 

Not 100% sold on that though, this might be fine but just wanted to raise it to see what you thought?

Reply:
Agreed. I've rewritten this to make the terminology clearer.